Bans the use of icons and icon_state strings as overlay items; Adds a note about abusing associated lists. (#58383)
tgstation/tgstation: .github/CONTRIBUTING.md

for (var/i in reagents)
```

### Icons are for image manipulation and defining an obj's `.icon` var, appearances are for everything else.

> Byond will allow you to use a raw icon file or even an icon datum for underlays, overlays, and what not (You can even use strings to refer to an icon state on the current icon). The issue is these get converted by byond to appearances on every overlay insert or removal involving them, and this process requires inserting the new appearance into the global list of appearances, and informing clients about them.

Reviewer: ```suggestion
BYOND will allow you to use a raw icon file or even an icon datum for underlays, overlays, and what not (you can even use strings to refer to an icon state on the current icon). The issue is these get converted by BYOND to appearances on every overlay insert or removal involving them, and this process requires inserting the new appearance into the global list of appearances, and informing clients about them.
```

Converting them yourself to appearances and storing this converted value will ensure this process only has to happen once for the lifetime of the round. Helper functions exist to do most of the work for you.

Bad:
```dm

> /obj/machine/update_overlays(var/blah)

Reviewer 2: ```suggestion
/obj/machine/update_overlays(blah)
```
No need for `var` boilerplate in the arguments. They can't be anything besides variables.
Same for the other entries.

	if (stat & broken)
		add_overlay(icon(broken_icon))  //this icon gets created, passed to byond, converted to an appearance, then deleted.
		return
	if (is_on)
		add_overlay("on") //also bad, the converstion to an appearance still has to happen
	else
		add_overlay(iconstate2appearance(icon, "off")) //this might seem alright, but not storing the value just moves the repeated appearance generation to this proc rather then the core overlay management. It would only be acceptable (and to some degree perferred) if this overlay is only ever added once (like in init code)
```

Good:
```dm
/obj/machine/update_overlays(var/blah)
	var/static/on_overlay = iconstate2appearance(icon, "on")
	var/static/off_overlay = iconstate2appearance(icon, "off")
	var/static/broken_overlay = icon2appearance(broken_icon)
	if (stat & broken)
		add_overlay(broken_overlay) 
		return
	if (is_on)
		add_overlay(on_overlay)
	else
		add_overlay(off_overlay)
	...
```

Note: images are appearances with extra steps, and don't incur the overhead in conversion.

> ### Do not abuse assoicated lists.

Reviewer: ```suggestion
### Do not abuse associated lists.
```

Associated lists that could instead be variables or statically defined number indexed lists will use more memory, as associated lists have a 24 bytes per item overhead (vs 8 for lists and most vars), and are slower to search compared to static/global variables and lists with known indexes.

Bad:
```dm
/obj/machine/update_overlays(var/blah)
	var/static/our_overlays = list("on" = iconstate2appearance(overlay_icon, "on"), "off" = iconstate2appearance(overlay_icon, "off"), "broken" = iconstate2appearance(overlay_icon, "broken"))
	if (stat & broken)
		add_overlay(our_overlays["broken"]) 
		return
	...
```

Good:
```dm
#define OUR_ON_OVERLAY 1
#define OUR_OFF_OVERLAY 2
#define OUR_BROKEN_OVERLAY 3
/obj/machine/update_overlays(var/blah)
	var/static/our_overlays = list(iconstate2appearance(overlay_icon, "on"), iconstate2appearance(overlay_icon, "off"), iconstate2appearance(overlay_icon, "broken"))
	if (stat & broken)
		add_overlay(our_overlays[OUR_BROKEN_OVERLAY])
		return
	...

#undef OUR_ON_OVERLAY
#undef OUR_OFF_OVERLAY
#undef OUR_BROKEN_OVERLAY
```

> Storing these in a flat (non-assoicated) list saves on memory, and using defines to reference locations in the list saves cpu time searching the list.

Reviewer: ```suggestion
Storing these in a flat (non-associated) list saves on memory, and using defines to reference locations in the list saves CPU time searching the list.
```

> Also Good:

Reviewer: ```suggestion
Also good:
```

```dm
/obj/machine/update_overlays(var/blah)
	var/static/on_overlay = iconstate2appearance(overlay_icon, "on")
	var/static/off_overlay = iconstate2appearance(overlay_icon, "off")
	var/static/broken_overlay = iconstate2appearance(overlay_icon, "broken")
	if (stat & broken)
		add_overlay(broken_overlay)
		return
	...
```

> Proc variables, static variables, and global variables are resolved at compile time, so the above is equalivent to the second example, but is easier to read, and avoids the need to store a list.

Reviewer: ```suggestion
Proc variables, static variables, and global variables are resolved at compile time, so the above is equivalent to the second example, but is easier to read, and avoids the need to store a list.
```

> Note: While there has historically been a strong impulse to use assoicated lists for caching of computed values, this is the easy way out and leaves a lot of hidden overhead. Please keep this in mind when designing core/root systems that are intended for use by other code/coders. Its normally better for consumers of such systems to handle their own caching using vars and number indexed lists, then for you to do it using assoicated lists.

Reviewer: ```suggestion
Note: While there has historically been a strong impulse to use associated lists for caching of computed values, this is the easy way out and leaves a lot of hidden overhead. Please keep this in mind when designing core/root systems that are intended for use by other code/coders. It's normally better for consumers of such systems to handle their own caching using vars and number indexed lists, than for you to do it using associated lists.
```

* The dlls section of tgs3.json is not designed for dlls that are purely `call()()`ed since those handles are closed between world reboots. Only put in dlls that may have to exist between world reboots.

> * Do not use icons, or icon_state strings as overlays. Byond (and by nature, our overlay management code) has to convert these to appearances every time they get added or removed from overlays, a costly operation. Instead Use `icon2appearance()` or `icon_state2appearance()`. These are global procs and can be used in static/global variable defines to avoid re-calling them with the same content.

Reviewer: ```suggestion
* Do not use icons, or icon_state strings as overlays. Byond (and by nature, our overlay management code) has to convert these to appearances every time they get added or removed from overlays, a costly operation. Instead, use `icon2appearance()` or `icon_state2appearance()`. These are global procs and can be used in static/global variable defines to avoid re-calling them with the same content.
```
